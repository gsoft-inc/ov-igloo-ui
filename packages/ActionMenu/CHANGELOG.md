# @igloo-ui/action-menu

## 1.3.2

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency
- Updated dependencies [a2df55d]
  - @igloo-ui/dropdown@1.6.4
  - @igloo-ui/list@0.5.2

## 1.3.1

### Patch Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.
- Updated dependencies [9739155]
  - @igloo-ui/dropdown@1.6.3
  - @igloo-ui/list@0.5.1

## 1.3.0

### Minor Changes

- 5402bcc: Created a "brand" toggle in the Toolbar and applied the new Workleap brand to some components. Also deprecated the Tag's appearance values and added new ones.

### Patch Changes

- Updated dependencies [5402bcc]
  - @igloo-ui/list@0.5.0

## 1.2.1

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.
- Updated dependencies [bf4ddef]
  - @igloo-ui/dropdown@1.6.1
  - @igloo-ui/list@0.4.3

## 1.2.0

### Minor Changes

- 5ed6021: Added the onAfterMenuClose callback to the action menu and dropdown components.

### Patch Changes

- Updated dependencies [5ed6021]
  - @igloo-ui/dropdown@1.5.0

## 1.1.4

### Patch Changes

- c3efaf5: Fixed tests and snapshots since react-aria generated ids changed and viewBox was added to icons.
- Updated dependencies [c3efaf5]
  - @igloo-ui/list@0.4.1

## 1.1.3

### Patch Changes

- 9d9d72a: Fixed WithinContainer story to show two cards and address z-index issues.
- Updated dependencies [9d9d72a]
  - @igloo-ui/dropdown@1.4.0

## 1.1.2

### Patch Changes

- Updated dependencies [9b707d1]
  - @igloo-ui/list@0.4.0

## 1.1.1

### Patch Changes

- 957387b: Other props for the action menu reference were overriding the onKeyDown event. Rearranged the order to fix this.
- Updated dependencies [0482ec6]
  - @igloo-ui/list@0.3.0

## 1.1.0

### Minor Changes

- 6e2e96a: Added the disablePortal prop for when you don't want the dropdown to be appended to the end of the body. False by default so it won't break the current behavior.

### Patch Changes

- Updated dependencies [6e2e96a]
- Updated dependencies [08a062e]
- Updated dependencies [6e2e96a]
  - @igloo-ui/dropdown@1.3.0
  - @igloo-ui/list@0.2.1

## 1.0.0

### Major Changes

- 4ff1942: Fixed some styling issues with the Action Menu. Moved two callbacks in action menu from ActionMenuProps to the props of the option.

  ## BREAKING CHANGE

  ActionMenu is called differently when using the callbacks for when an option is clicked or determining if the menu is closed on select. They were removed from ActionMenuProps and added to the option props instead.

### Patch Changes

- Updated dependencies [4ff1942]
- Updated dependencies [0b8a62f]
  - @igloo-ui/list@0.2.0
  - @igloo-ui/dropdown@1.2.0

## 0.1.2

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.
- Updated dependencies [a41e1d4]
  - @igloo-ui/dropdown@1.1.4
  - @igloo-ui/list@0.1.3

## 0.1.1

### Patch Changes

- bf478b5: Added missing tests. Fixed missing code in stories when clicking 'Show code'. Updated Dropdown mock. Added new optional prop highlightToday to help with test snapshots.
- Updated dependencies [bf478b5]
  - @igloo-ui/dropdown@1.1.3

## 0.1.0

### Minor Changes

- 3f50827: Feat: The initial release of the action menu component

### Patch Changes

- Updated dependencies [3f50827]
  - @igloo-ui/dropdown@1.1.2
