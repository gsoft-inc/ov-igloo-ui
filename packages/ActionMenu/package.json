{
  "name": "@igloo-ui/action-menu",
  "version": "1.5.0",
  "main": "dist/ActionMenu.js",
  "module": "dist/ActionMenu.js",
  "types": "dist/ActionMenu.d.ts",
  "styles": "dist/action-menu.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "@igloo-ui/dropdown": "^1.8.0",
    "@igloo-ui/list": "^0.6.0",
    "@hopper-ui/tokens": "^3.1.3",
    "classnames": "^2.3.2"
  },
  "devDependencies": {
    "@igloo-ui/icons": "^1.12.2",
    "@igloo-ui/button": "^0.8.0",
    "@igloo-ui/helper-text": "^0.3.0",
    "@igloo-ui/icon-button": "^1.2.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
