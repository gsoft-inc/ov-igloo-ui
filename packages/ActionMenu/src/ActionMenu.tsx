import * as React from "react";
import cx from "classnames";

import List, { type OptionType, type Option } from "@igloo-ui/list";
import Dropdown, { type Position } from "@igloo-ui/dropdown";

import "./action-menu.scss";

export enum Keys {
    Enter = "Enter",
    Space = " ",
    ArrowDown = "ArrowDown",
    ArrowUp = "ArrowUp",
    Escape = "Escape",
    Tab = "Tab",
    Home = "Home",
    End = "End"
}

export type FocusDirection = "first" | "last" | "up" | "down";

export interface ActionMenuOption extends Omit<Option, "type"> {
    /** Whether or not the action menu should close when an option is selected */
    closeOnSelect?: boolean | ((option: OptionType) => boolean);
    /** Callback when an option is selected */
    onClick?: () => void;
}

export interface ActionMenuProps extends React.ComponentProps<"div"> {
    /** Add a specific class to the action menu */
    className?: string;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** Whether or not the menu should use ReactPortal
   * to append to the body */
    disablePortal?: boolean;
    /** Add a specific class to the dropdown */
    dropdownClassName?: string;
    /** Offset of the dropdown */
    dropdownOffset?: number;
    /** Whether or not the action menu should be open by default */
    isOpen?: boolean;
    /** Callback when the action menu is closed and animations are done */
    onAfterMenuClose?: () => void;
    /** Callback when the action menu is closed  */
    onMenuClose?: () => void;
    /** Callback when the action menu is opened  */
    onMenuOpen?: () => void;
    /** A list of options to display in the action menu */
    options: ActionMenuOption[];
    /** Position of the action menu */
    position?: Position;
    /** Render the reference element to be able to add the
     * reference props directly */
    renderReference: (
        props: React.HTMLProps<HTMLButtonElement>,
    ) => React.ReactElement;
    /** Footer to display helper text or other content below the options */
    footer?: React.ReactNode;
}

const ActionMenu: React.FunctionComponent<ActionMenuProps> = ({
    className,
    dataTest,
    disablePortal = false,
    dropdownClassName,
    dropdownOffset,
    isOpen = false,
    onAfterMenuClose,
    onMenuClose,
    onMenuOpen,
    options,
    position = "bottom-end",
    renderReference,
    footer,
    ...rest
}: ActionMenuProps) => {
    const actionMenuOptions = options.map((option): OptionType => {
        return {
            ...option,
            type: "list"
        };
    });

    const [showMenu, setShowMenu] = React.useState(isOpen);
    const [currentFocusedOption, setCurrentFocusedOption] =
        React.useState<OptionType>();

    const isOptionDisabled = (option: OptionType | undefined): boolean => {
        if (option?.type === "list") {
            return option?.disabled ?? false;
        }

        return false;
    };

    const toggleMenu = (open: boolean): void => {
        setShowMenu(open);

        if (!open) {
            setCurrentFocusedOption(undefined);
            if (onMenuClose) {
                onMenuClose();
            }
        } else if (onMenuOpen) {
            onMenuOpen();
        }
    };

    const closeMenuOnSelect = (option: OptionType): boolean => {
        const actionMenuOption = options.find(
            actionMenuOption => actionMenuOption.value === option.value
        );
        const closeOnSelect = actionMenuOption?.closeOnSelect ?? true;
        if (typeof closeOnSelect === "function") {
            return closeOnSelect(option);
        }

        return closeOnSelect;
    };

    const selectOption = (option: OptionType): void => {
        const actionMenuOption = options.find(
            actionMenuOption => actionMenuOption.value === option.value
        );
        const onOptionSelect = actionMenuOption?.onClick;
        if (onOptionSelect) {
            onOptionSelect();
        }

        if (closeMenuOnSelect(option)) {
            toggleMenu(false);
        }
    };

    const hoverOption = (option: OptionType): void => {
        setCurrentFocusedOption(option);
    };

    const focusOption = (direction: FocusDirection = "first"): void => {
        const enabledOptions = actionMenuOptions.filter(
            option => !isOptionDisabled(option)
        );
        if (!enabledOptions.length) { return; }

        let currentFocusedIndex = -1;
        if (currentFocusedOption) {
            currentFocusedIndex = enabledOptions.findIndex(
                enabledOption => enabledOption.value === currentFocusedOption.value
            );
        }

        switch (direction) {
            case "up":
                setCurrentFocusedOption(
                    enabledOptions[
                        currentFocusedIndex > 0
                            ? currentFocusedIndex - 1
                            : enabledOptions.length - 1
                    ]
                );
                break;
            case "down":
                setCurrentFocusedOption(
                    enabledOptions[(currentFocusedIndex + 1) % enabledOptions.length]
                );
                break;
            case "last":
                setCurrentFocusedOption(enabledOptions[enabledOptions.length - 1]);
                break;
            default:
                setCurrentFocusedOption(enabledOptions[0]);
                break;
        }
    };

    const handleOnKeyDown = (
        keyboardEvent: React.KeyboardEvent<HTMLButtonElement>
    ): void => {
        switch (keyboardEvent.key) {
            case Keys.Escape:
                if (showMenu) {
                    toggleMenu(showMenu);
                }
                break;
            case Keys.Enter:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();
                if (currentFocusedOption) {
                    selectOption(currentFocusedOption);
                }
                if ((!currentFocusedOption && showMenu) || !showMenu) {
                    toggleMenu(!showMenu);
                }
                break;
            case Keys.Space:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();
                if (!showMenu) {
                    toggleMenu(true);
                }
                break;
            case Keys.ArrowUp:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();

                focusOption("up");
                break;
            case Keys.ArrowDown:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();
                focusOption("down");
                break;
            case Keys.Home:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();

                focusOption("first");
                break;
            case Keys.End:
                keyboardEvent.preventDefault();
                keyboardEvent.stopPropagation();

                focusOption("last");
                break;
            case Keys.Tab:
                if (showMenu) {
                    toggleMenu(showMenu);
                }

                break;
            default:
                break;
        }
    };

    const actionMenuClassname = cx("ids-action-menu", className);

    return (
        <div className={actionMenuClassname} data-test={dataTest} {...rest}>
            <Dropdown
                key="selectDropdown"
                role="menu"
                disablePortal={disablePortal}
                content={
                    <List
                        options={actionMenuOptions}
                        onOptionFocus={hoverOption}
                        onOptionBlur={() => setCurrentFocusedOption(undefined)}
                        onOptionChange={selectOption}
                        focusedOption={currentFocusedOption}
                        disableTabbing
                    />
                }
                isOpen={showMenu}
                className={cx("ids-action-menu__dropdown", dropdownClassName)}
                dropdownOffset={dropdownOffset}
                position={position}
                onClose={() => toggleMenu(false)}
                onAfterClose={onAfterMenuClose}
                renderReference={(refProps: React.HTMLProps<HTMLButtonElement>) => {
                    return renderReference({
                        ...refProps,
                        onClick: () => toggleMenu(!showMenu),
                        onKeyDown: e => {
                            handleOnKeyDown(e);
                        },
                        className: "ids-action-menu__trigger"
                    });
                }}
                footer={footer}
            />
        </div>
    );
};

export default ActionMenu;
