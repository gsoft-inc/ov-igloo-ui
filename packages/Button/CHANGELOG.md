# Change Log

## 0.8.3

### Patch Changes

- 97edb0e: The button is now square when there's only an icon showing on mobile.

## 0.8.2

### Patch Changes

- 6ea531f: Updated dependency versions for react-aria packages, luxon and hopper

## 0.8.1

### Patch Changes

- f30fe4b: Color adjustments for ghost buttons

## 0.8.0

### Minor Changes

- 4846c59: Wrapped Igloo CSS in @layer to isolate it.

## 0.7.7

### Patch Changes

- cfb5619: Fixed button padding. Also removed styles on loader buttons so they don't show up on focus.
- cfb5619: Updated hopper token versions and button versions.

## 0.7.6

### Patch Changes

- 5c12940: Fixed button padding. Also removed styles on loader buttons so they don't show up on focus.

## 0.7.5

### Patch Changes

- 37260f9: Updated Button to use :focus-visible and not :focus. The old method isn't needed anymore since we no longer support IE. Also added missing line-height a ensured box-sixing is applied to all children of button.

## 0.7.4

### Patch Changes

- 7d36815: Updated hopper token versions and igloo icon versions.

## 0.7.3

### Patch Changes

- a2df55d: Updated to latest Hopper Token dependency

## 0.7.2

### Patch Changes

- 592afb0: change the border value to render on base 10 and 16.

## 0.7.1

### Patch Changes

- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.

## 0.7.0

### Minor Changes

- 8eeac98: Rebrand of dialog and modal / media query fix in rems

## 0.6.0

### Minor Changes

- 5402bcc: Created a "brand" toggle in the Toolbar and applied the new Workleap brand to some components. Also deprecated the Tag's appearance values and added new ones.

## 0.5.3

### Patch Changes

- 7662390: Removed the :active styles for Button since they were not needed or wanted. .ids-btn--active remains the same.

## 0.5.2

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.

## 0.5.1

### Patch Changes

- c3efaf5: Fixed tests and snapshots since react-aria generated ids changed and viewBox was added to icons.

## 0.5.0

### Minor Changes

- 357133d: Add new props for reset the size to the default on mobile

## 0.4.2

### Patch Changes

- 949db42: Added a display name to the Button so that it displays properly in storybook's code snippet.

## 0.4.1

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.

## 0.4.0

### Minor Changes

- d2e08a0: Feat: Added the ghost secondary appearance of the button.

### Patch Changes

- d5c91f5: link shared components with aliases

## 0.3.7

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.3.6

### Patch Changes

- 0640409: Update @igloo-ui/icons to 1.6.0
- 0640409: udpate classnames to 2.3.2

## 0.3.5

### Patch Changes

- d971143: All tests were updated to use testing library and not enzyme

## 0.3.4

### Patch Changes

- 4bd97f4: fix the hover state rendering of the disabled secondary button

## 0.3.3

### Patch Changes

- 8bc7f2c: add not-allowed cursor on disabled state

## 0.3.2

### Patch Changes

- f1611fd: Temporary quick fix on ghost danger button by adding `.ids-btn` prefix

## 0.3.1

### Patch Changes

- 4f79069: enforce button ghost danger ui

## 0.3.0

### Minor Changes

- 4de58ec: Add new appearance Ghost Danger in `Button`

### Patch Changes

- 38a4994: Update dependency on @igloo-ui/icons and other @igloo libs

## 0.2.3

### Patch Changes

- c89877e: Fix render of ghost button in ie11

## 0.2.2

### Patch Changes

- 0a86f13: Update content of README

## 0.2.1

### Patch Changes

- 3eb651d: fix the width of the button as link

## 0.2.0

### Minor Changes

- f019f88: Add the ability to display a link by an `a` tag

## 0.1.12

### Patch Changes

- ac5d81a: add forwardRef

## 0.1.11

### Patch Changes

- 8c8686e: Change @import of font.css file for a @use in SASS files

## 0.1.10

### Patch Changes

- 6d91ab6: Fix secondary button empty background on hover state

## 0.1.9

### Patch Changes

- 4a256e3: IE11 compatibility

## 0.1.8

### Patch Changes

- 1e1c6e5: Chore Update packages

## 0.1.7

### Patch Changes

- ae0a0a5: Fixed the case where the content of the Button is not text, but that we tried to display the object in the title property.

## 0.1.6

### Patch Changes

- 0df8530: Rename Props to ButtonProps

## 0.1.5

### Patch Changes

- 8450a04: Fix resizing effect when loading is complete

## 0.1.4

### Patch Changes

- 2da8010: Fix render of elements with the css var --grey-{...}

## 0.1.3

### Patch Changes

- 31593ad: Change Premium button color and focus, how buttons handle overflow, how active button handle hover and some tweaks to the dev tooling on Windows.

## 0.1.2

### Patch Changes

- c290c9a: remove `src` folder from the package

## 0.1.1

### Patch Changes

- 4ed4836: Add prepublish-ci script for build lib

## 0.1.0

### Minor Changes

- 76a63cb: - **button:** add loading effect ([2a8d84f](https://github.com/gsoft-inc/ov-igloo-ui/commit/2a8d84fa8739f170e0189d053ec4f5136ad3c4d4))
  - **button:** add of the responsive view ([364a926](https://github.com/gsoft-inc/ov-igloo-ui/commit/364a92611aae51ab555feaa6e503933ca80f1433))
  - **button:** classname for OV standard compoliance ([8ed7afb](https://github.com/gsoft-inc/ov-igloo-ui/commit/8ed7afb047f552f52fd151ef2da04be49f9908f1))
  - **button:** change the css variables name ([68f710d](https://github.com/gsoft-inc/ov-igloo-ui/commit/68f710d16e653681ff90e9124ebbafd759dfe057))
  - add declaration for `renderContent`  ([5827120](https://github.com/gsoft-inc/ov-igloo-ui/commit/58271200d14a3372dc81fd1d9189e5dfade87143))
  - add focus state ([31910e4](https://github.com/gsoft-inc/ov-igloo-ui/commit/31910e48ff15d197323666bae376116d6fe97c56))
  - remove focus when the button is clicked ([caa6d45](https://github.com/gsoft-inc/ov-igloo-ui/commit/caa6d4544979d605794b23b64fdd838af3c61a23))
  - update stylelint for OV standard compliance ([184d006](https://github.com/gsoft-inc/ov-igloo-ui/commit/184d006ecb10d8b0a19a2303f7399b54b895e992))
  - configure stylelint ([d27df67](https://github.com/gsoft-inc/ov-igloo-ui/commit/d27df672a99145eb3a134cab4549b0e0998a6d0f))

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0-alpha.0 (2021-10-29)

### Bug Fixes

- add declaration for `renderContent`  ([5827120](https://github.com/gsoft-inc/ov-igloo-ui/commit/58271200d14a3372dc81fd1d9189e5dfade87143))
- **button:** classname for OV standard compoliance ([8ed7afb](https://github.com/gsoft-inc/ov-igloo-ui/commit/8ed7afb047f552f52fd151ef2da04be49f9908f1))
- remove focus when the button is clicked ([caa6d45](https://github.com/gsoft-inc/ov-igloo-ui/commit/caa6d4544979d605794b23b64fdd838af3c61a23))
- configure stylelint ([d27df67](https://github.com/gsoft-inc/ov-igloo-ui/commit/d27df672a99145eb3a134cab4549b0e0998a6d0f))
- update stylelint for OV standard compliance ([184d006](https://github.com/gsoft-inc/ov-igloo-ui/commit/184d006ecb10d8b0a19a2303f7399b54b895e992))

### Features

- add focus state ([31910e4](https://github.com/gsoft-inc/ov-igloo-ui/commit/31910e48ff15d197323666bae376116d6fe97c56))
- **button:** add loading effect ([2a8d84f](https://github.com/gsoft-inc/ov-igloo-ui/commit/2a8d84fa8739f170e0189d053ec4f5136ad3c4d4))
- **button:** add of the responsive view ([364a926](https://github.com/gsoft-inc/ov-igloo-ui/commit/364a92611aae51ab555feaa6e503933ca80f1433))
- **button:** change the css variables name ([68f710d](https://github.com/gsoft-inc/ov-igloo-ui/commit/68f710d16e653681ff90e9124ebbafd759dfe057))
