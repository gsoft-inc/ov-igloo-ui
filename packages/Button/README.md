# Button

Buttons are used primarily for actions.

<Example />

<ReferenceLinks />

## Installation

Install the component from your command line.

```bash
npm install @igloo-ui/button
# or with yarn
yarn add @igloo-ui/button
```

## Usage

Then to use the component in your code just import it!

```jsx
import { Button } from '@igloo-ui/button';

<Button>Primary</Button>;
```
