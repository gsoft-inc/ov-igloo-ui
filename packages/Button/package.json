{
  "name": "@igloo-ui/button",
  "version": "0.2.3",
  "main": "dist/Button.js",
  "module": "dist/Button.js",
  "types": "dist/Button.d.ts",
  "styles": "dist/button.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Franck franck.gaudin@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "@igloo-ui/tokens": "^1.1.0",
    "classnames": "^2.3.1"
  },
  "devDependencies": {
    "@igloo-ui/icons": "^0.6.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
