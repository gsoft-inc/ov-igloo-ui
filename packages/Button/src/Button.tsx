import * as React from 'react';
import cx from 'classnames';

import './button.scss';

export type Appearance =
  | 'primary'
  | 'secondary'
  | 'premium'
  | 'ghost'
  | 'danger'
  | { type: 'ghost'; variant?: 'secondary' }
  | { type: 'ghost'; variant?: 'danger' };
export type Size = 'small' | 'medium';

type AsProp<C extends React.ElementType> = {
  as?: C;
};

type PropsToOmit<C extends React.ElementType, P> = keyof (AsProp<C> & P);

type PolymorphicComponentProp<
  C extends React.ElementType,
  Props = {}
> = React.PropsWithChildren<Props & AsProp<C>> &
  Omit<React.ComponentPropsWithoutRef<C>, PropsToOmit<C, Props>>;

type PolymorphicComponentPropWithRef<
  C extends React.ElementType,
  Props = {}
> = PolymorphicComponentProp<C, Props> & { ref?: PolymorphicRef<C> };

type PolymorphicRef<C extends React.ElementType> =
  React.ComponentPropsWithRef<C>['ref'];

export interface ButtonOwnProps {
  /** The content to display inside the button */
  children?: React.ReactNode;
  /** Disabled the button, the user cannot click on them */
  disabled?: boolean;
  /** Sets the button in an active state */
  active?: boolean;
  /** Replaces button text with a spinner while a
   * background action is being performed */
  loading?: boolean;
  /** Changes the size of button, giving more or less padding */
  size?: Size;
  /** Button appearance */
  appearance?: Appearance;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Icon to display to the left of button content */
  iconLeading?: React.ReactNode;
  /** Icon to display to the right of button content */
  iconTrailing?: React.ReactNode;
  /** Display only the icon in mobile */
  showOnlyIconOnMobile?: boolean;
  /** Callback when clicked */
  onClick?: () => void;
  /** Optional prop to specify the type of the Button */
  type?: 'button' | 'reset' | 'submit';
  /** Add a data-intercom-target with unique id to link a
   * component to a Product Tour step */
  intercomTarget?: string;
  /** Add a specific class to the button */
  className?: string;
  //** Return to the default size on mobile */
  resetSizeOnMobile?: boolean;
}

type ButtonProps<C extends React.ElementType> = PolymorphicComponentPropWithRef<
  C,
  ButtonOwnProps
>;

type ButtonComponent<C extends React.ElementType = 'button'> = {
  <T extends React.ElementType = C>(
    props: ButtonProps<T>
  ): React.ReactElement | null;

  displayName?: string;
};

const Button: ButtonComponent = React.forwardRef(
  <C extends React.ElementType = 'button'>(
    {
      children,
      disabled = false,
      active = false,
      loading = false,
      size = 'medium',
      appearance = 'primary',
      type = 'button',
      onClick,
      dataTest,
      iconLeading,
      iconTrailing,
      showOnlyIconOnMobile,
      resetSizeOnMobile = false,
      intercomTarget,
      className,
      as,
      ...rest
    }: ButtonProps<C>,
    ref?: PolymorphicRef<C>
  ) => {
    const hasIconLeading = iconLeading !== undefined;
    const hasIconTrailing = iconTrailing !== undefined;
    const hasIcon = hasIconLeading || hasIconTrailing;
    const childrenIsAString =
      typeof children === 'string' || children instanceof String;

    const normalizeAppearanceClass = (appearance: Appearance): {} => {
      if (typeof appearance === 'object') {
        return {
          [`ids-btn--${appearance.type}`]: !appearance.variant,
          [`ids-btn--${appearance.type}-${appearance.variant}`]:
            appearance.variant,
        };
      }

      return {
        [`ids-btn--${appearance}`]: appearance !== 'primary',
      };
    };

    const appearanceClasses = normalizeAppearanceClass(appearance);

    const classes = cx('ids-btn', className, appearanceClasses, {
      'ids-btn--small': size === 'small',
      'ids-btn--active': active,
      'ids-btn--loading': loading,
      'ids-btn--disabled': as === 'a' && disabled,
      'ids-btn--mobile': showOnlyIconOnMobile,
      'has-icon': hasIcon,
      'has-icon--leading': hasIconLeading,
      'has-icon--trailing': hasIconTrailing,
      'ids-btn--size-mobile': resetSizeOnMobile,
    });

    const renderContent = (): JSX.Element => {
      return (
        <>
          {hasIconLeading && iconLeading}
          {showOnlyIconOnMobile ? (
            <span
              className={cx('ids-btn__label', {
                'is-hidden': loading,
              })}
            >
              {children}
            </span>
          ) : (
            <span
              className={cx({
                'is-hidden': loading,
              })}
            >
              {children}
            </span>
          )}
          {hasIconTrailing && iconTrailing}
        </>
      );
    };

    const Component = as || 'button';

    return (
      <Component
        disabled={disabled}
        className={classes}
        data-test={dataTest}
        data-intercom-target={intercomTarget}
        type={type}
        title={childrenIsAString ? children?.toString() : ''}
        onClick={onClick}
        ref={ref}
        {...rest}
      >
        {loading && <div className="ids-loader" />}
        {renderContent()}
      </Component>
    );
  }
);

Button.displayName = 'Button'; // This fixed issue OV-42946

export default Button;
