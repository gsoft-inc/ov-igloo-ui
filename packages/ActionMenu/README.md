# ActionMenu

This component is used to display a menu that will contain a list of menu options. The menu is dismissible if you click outside the dropdown.

<Example is="custom" />
<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/action-menu` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/action-menu
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/action-menu
```

## Usage

Then to use the component in your code just import it!

```jsx
import ActionMenu from '@igloo-ui/action-menu';
import Button from '@igloo-ui/button';

<ActionMenu
  options={[
    {
      label: 'Add Item',
      value: 'add',
    },
    {
      label: 'Delete Item',
      value: 'delete',
      disabled: true,
    },
    {
      label: 'Copy Item',
      value: 'copy',
    },
  ]}
  renderReference={(refProps) => {
    return (
      <Button appearance="secondary" {...refProps}>
        Button
      </Button>
    );
  }}
/>;
```
