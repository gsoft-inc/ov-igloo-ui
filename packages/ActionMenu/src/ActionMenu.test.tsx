/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import MockDropdown from '@igloo-ui/dropdown/src/__mocks__/Dropdown.mock';
import Button from '@igloo-ui/button';
import ActionMenu, { ActionMenuOption, ActionMenuProps } from './ActionMenu';

const actionMenuList: ActionMenuOption[] = [
  {
    label: 'Add Item',
    value: 'add',
  },
  {
    label: 'Delete Item',
    value: 'delete',
    disabled: true,
  },
  {
    label: 'Copy Item',
    value: 'copy',
  },
];

const actionMenuProps: ActionMenuProps = {
  isOpen: true,
  dataTest: 'ids-action-menu',
  options: actionMenuList,
  renderReference: (refProps: any) => {
    return (
      <Button appearance="secondary" {...refProps}>
        Button
      </Button>
    );
  },
};

const setup = (props = actionMenuProps) => {
  return render(<ActionMenu {...props} />);
};

jest.mock('@igloo-ui/dropdown', () => ({
  __esModule: true,
  default: jest.fn(MockDropdown),
}));

describe('ActionMenu', () => {
  test('It should render without errors and a snapshot', () => {
    const { asFragment } = setup();
    const wrapper = screen.getByTestId('ids-action-menu');
    expect(wrapper).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It renders an option', () => {
    setup();
    const listItem = screen.getByText('Add Item');
    expect(listItem).toBeInTheDocument();
  });

  test('It calls the onClick of the option', () => {
    let selected = false;
    setup({
      ...actionMenuProps, options: [
        {
          label: 'Add Item',
          value: 'add',
          onClick: () => {
            selected = true;
          }
        },
      ]
    });
    const listItem = screen.getByText('Add Item');
    if (listItem) {
      fireEvent.click(listItem);
    }
    expect(selected).toBeTruthy();
  });
});
