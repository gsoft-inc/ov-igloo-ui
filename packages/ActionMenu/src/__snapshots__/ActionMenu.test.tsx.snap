// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMenu It should render without errors and a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-action-menu"
    data-test="ids-action-menu"
  >
    <button
      class="ids-btn ids-action-menu__trigger ids-btn--secondary"
      title="Button"
      type="button"
    >
      <span
        class=""
      >
        Button
      </span>
    </button>
    <div
      class="ids-dropdown ids-action-menu__dropdown ids-dropdown--bottom-end"
      data-show="true"
      style="position: fixed; top: 0px; left: 0px;"
    >
      <div
        class="ids-dropdown__scroll-content"
      >
        <ul
          class="ids-list ids-list--compact"
          role="listbox"
          tabindex="-1"
        >
          <li
            aria-selected="false"
            class="ids-list-item ids-list-item--compact ids-list-item--selectable"
            role="option"
            tabindex="-1"
          >
            <div
              class="ids-list-item__content"
            >
              <span
                class="ids-list-item__text"
              >
                <span
                  class="ids-list-item__text-body"
                >
                  <span
                    class="ids-list-item__text-content"
                  >
                    <span
                      class="ids-list-item__text-label"
                    >
                      Add Item
                    </span>
                  </span>
                </span>
              </span>
            </div>
          </li>
          <li
            aria-selected="false"
            class="ids-list-item ids-list-item--compact ids-list-item--disabled"
            role="option"
            tabindex="-1"
          >
            <div
              class="ids-list-item__content"
            >
              <span
                class="ids-list-item__text"
              >
                <span
                  class="ids-list-item__text-body"
                >
                  <span
                    class="ids-list-item__text-content"
                  >
                    <span
                      class="ids-list-item__text-label"
                    >
                      Delete Item
                    </span>
                  </span>
                </span>
              </span>
            </div>
          </li>
          <li
            aria-selected="false"
            class="ids-list-item ids-list-item--compact ids-list-item--selectable"
            role="option"
            tabindex="-1"
          >
            <div
              class="ids-list-item__content"
            >
              <span
                class="ids-list-item__text"
              >
                <span
                  class="ids-list-item__text-body"
                >
                  <span
                    class="ids-list-item__text-content"
                  >
                    <span
                      class="ids-list-item__text-label"
                    >
                      Copy Item
                    </span>
                  </span>
                </span>
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
